### PR TITLE
Implement Request Funnel dashboards and fix multi-message bug

### PR DESCRIPTION
**Pull request by SavvasM1**
Adds a "PR Review Support Automation" funnel to the Tenant Requests dashboard, and fixes a bug where custom PR-detection override messages were posted once per PR.

**Funnel:**

Replaces the single "Intervention Rate" card with a 3-step funnel, all counted by ticket-creation date so each is a subset of the previous:
- Total Requests — all support tickets in the window.
- Handled by Bot — % that are PR tickets (≥1 tracked PR).
- Needed Manual Escalation — % of those PR tickets a human manually escalated.

**Bug fix — Multiple messages with custom override**

A repo with a custom override message got one message per PR when several PRs for the same repo arrived in a single message. Now the override posts once per repo/notification-type group, matching the default grouped message. Applies to no-SLA and SLA-backed repos. A debug log records the grouped PR numbers.

**Follow-up**

Custom override messages still can't list all grouped PRs (templating exposes a single PR) — tracked separately for the detected event.

Dashboards:

image

### CODE REVIEW

**Review by jescard-tamer-cecg — COMMENTED** [review]
## Automated code review — max effort

Reviewed across line-by-line, removed-behavior, cross-file, language pitfalls, SQL correctness, reuse/simplification/efficiency, altitude, and conventions angles, plus a gap sweep. Net: **1 should-block correctness issue**, 2 medium robustness/correctness issues, and several cleanup / efficiency / test-gap items — all left as inline comments below.

**One finding can't be anchored inline (file unchanged):** the route-registration guard test (`TenantInsightsRouteRegistrationTest`) wasn't extended to `/request-breakdown`. Its sibling endpoints are covered for the enabled/disabled feature-gating; the new endpoint isn't — so a future gating regression (404 when the feature is enabled, or data served when it's off) would slip past the very suite meant to catch it.

Good PR overall: the funnel's nested-subset invariant genuinely holds, bind ordering is correct, and test coverage is broad.

_Posted by an automated review pass — treat as suggestions._

File: api/service/src/main/java/com/coreeng/supportbot/prtracking/PrDetectionService.java [hunk 1]
```diff
@@ -824,7 +827,17 @@ private void postGroupedNotifications(List<PendingNotification> repoNotifs, Mess
...
+                // rendered with a single PR's context, so the other PRs can't be named in it — log
+                // them so the collapse is observable (e.g. if a repo uses a PR-specific override).
+                log.atDebug()
+                        .addArgument(repo)
+                        .addArgument(() -> group.stream()
+                                .map(n -> String.valueOf(n.prNumber()))
+                                .collect(Collectors.joining(", ")))
+                        .log(
+                                "Custom override for repo {} posted once for PRs {} (rendered with the first PR's context)");
+                postSingleNotification(group.getFirst(), queryTs, channelId);
```
Comment (jescard-tamer-cecg) [inline api/service/src/main/java/com/coreeng/supportbot/prtracking/PrDetectionService.java:902]: **Should-block: custom PR-specific override messages now silently drop PRs 2..N.** This collapses the override group to `group.getFirst()`, rendered with only the first PR's context. The comment above states the override text is repo/event-level (not PR-specific) — but that premise doesn't hold: the renderer exposes `pr_number`/`pr_url`, and the project's own docs ship per-event overrides built with `string(pr_number)` (`api/service/docs/configuration.md:711-716`, `application.yaml:170-180`). So a repo using a documented per-event override that receives 2+ PRs in one poll now posts a single message naming only the first PR; the rest survive only in a debug log. The default (non-override) path just below lists *every* PR via `prList`, so this branch is now strictly worse than the per-PR behavior it replaced. Consider keeping per-PR posting when the template references `pr_number`, or adding a `pr_list`-style var so the override can name all grouped PRs.

File: api/service/src/main/java/com/coreeng/supportbot/prtracking/JdbcPrTrackingRepository.java [hunk 2]
```diff
@@ -376,6 +376,51 @@ WHERE EXISTS (
...
+        // Compares PR tickets against ALL support tickets that came in during the window, so every count
+        // is bucketed by when the support request arrived — making the PR and intervention counts true subsets of the
+        // total.
+        // Bind order matches the textual '?' order: the manual-source filter (SELECT list) first,
+        // then the query.date range in the WHERE clause.
+        List<Object> binds = new ArrayList<>();
+        binds.add(EscalationSource.manual.name());
+        StringBuilder dateFilter = new StringBuilder();
+        if (dateFrom != null) {
+            dateFilter.append("AND q.date::date >= ?::date ");
```
Comment (jescard-tamer-cecg) [inline api/service/src/main/java/com/coreeng/supportbot/prtracking/JdbcPrTrackingRepository.java]: **`q.date::date` cast hurts both correctness and performance.** `q.date` is `timestamptz` with `query_date_idx` on it (`V1__init.sql:26`). (1) Casting to `::date` truncates in the **DB session timezone**, so a ticket created near midnight at a window edge can fall into the adjacent day — this also makes the exact-count functional tests flaky on non-UTC CI / near local midnight (they derive boundaries from `LocalDate.now()` in the JVM zone). (2) The cast makes the predicate **non-sargable**, so `query_date_idx` can't be used and this dashboard query full-scans `ticket ⋈ query`. A half-open range fixes both: `q.date >= ?::date AND q.date < ?::date + interval '1 day'`. (Pre-existing pattern in `buildDateFilter` on `pr_created_at`, but it now drives a user-facing funnel.)

File: api/service/src/main/java/com/coreeng/supportbot/prtracking/RequestBreakdown.java [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+ * had a PR, and how many of the PR tickets needed a manual engineer escalation.
+ *
+ * <p>All three counts share one date anchor — ticket creation (the {@code query} table's
+ * {@code date} column) — so they are nested subsets: {@code interventionPrTickets <=
+ * totalPrTickets <= totalSupportTickets}. The
+ * invariant is enforced here so a query regression surfaces loudly rather than rendering a
+ * nonsensical >100% rate in the UI.
+ */
+public record RequestBreakdown(long totalSupportTickets, long totalPrTickets, long interventionPrTickets) {
+    public RequestBreakdown {
```
Comment (jescard-tamer-cecg) [inline api/service/src/main/java/com/coreeng/supportbot/prtracking/RequestBreakdown.java:14]: **A throw here 500s the whole stats dashboard.** There's no `@ControllerAdvice` anywhere in `api/service`, so an `IllegalArgumentException` from this compact constructor surfaces as HTTP 500 — and the UI does `error = statsError || breakdownError`, which blanks the *entire* repo table (`tenant-requests.tsx`), not just the funnel cards (which already degrade to a dash). The invariant is actually unreachable from the current SQL (all three counts are `COUNT(DISTINCT t.id)` over the same rows with progressively stricter FILTERs), so this can't fire today — but as a guard against a future query regression the blast radius is wrong. Prefer clamping + a warning log (or letting the UI dash) over a fatal throw at the data layer.

File: ui/src/lib/hooks/index.ts [hunk 4]
```diff
@@ -513,6 +514,15 @@ export function useEscalationBreakdown(dateFrom?: string, dateTo?: string, enabl
   });
 }
 
+export function useRequestBreakdown(dateFrom?: string, dateTo?: string, enabled = true) {
```
Comment (jescard-tamer-cecg) [inline ui/src/lib/hooks/index.ts:507]: **Dead code left behind by this swap.** Now that `tenant-requests.tsx` uses `useRequestBreakdown`, the old `useEscalationBreakdown` (line 508), the `EscalationBreakdown` type (`lib/types/dashboard.ts:129`), and the `app/api/tenant-insights/escalation-breakdown/route.ts` proxy have **zero UI consumers** (confirmed by grep). Worth removing them in this PR — and deciding whether the backend `escalationBreakdown` endpoint is still wanted, since it anchors on `pr_created_at` while this new funnel anchors on `q.date`, so the two definitions of a 'PR ticket' can silently drift.

File: api/service/src/main/java/com/coreeng/supportbot/prtracking/JdbcPrTrackingRepository.java [hunk 5]
```diff
@@ -376,6 +376,51 @@ WHERE EXISTS (
...
+    @Override
+    public RequestBreakdown getRequestBreakdown(@Nullable LocalDate dateFrom, @Nullable LocalDate dateTo) {
+        // Compares PR tickets against ALL support tickets that came in during the window, so every count
+        // is bucketed by when the support request arrived — making the PR and intervention counts true subsets of the
+        // total.
+        // Bind order matches the textual '?' order: the manual-source filter (SELECT list) first,
+        // then the query.date range in the WHERE clause.
+        List<Object> binds = new ArrayList<>();
+        binds.add(EscalationSource.manual.name());
+        StringBuilder dateFilter = new StringBuilder();
```
Comment (jescard-tamer-cecg) [inline api/service/src/main/java/com/coreeng/supportbot/prtracking/JdbcPrTrackingRepository.java]: **Reuse `buildDateFilter` rather than re-inlining.** `buildDateFilter` (used by `getEscalationBreakdown` and `getRepoInsights`) already implements this StringBuilder + binds pattern; it just hardcodes `pr_created_at`. Parameterizing it by column (e.g. `buildDateFilter('q.date', dateFrom, dateTo, binds)`) would let all three call sites share one implementation, so a future date-handling fix can't silently reach only two of them.

File: api/service/src/main/java/com/coreeng/supportbot/prtracking/rest/TenantInsightsController.java [hunk 6]
```diff
@@ -86,6 +86,16 @@ public EscalationBreakdown escalationBreakdown(
         return prTrackingRepository.getEscalationBreakdown(dateFrom, dateTo);
     }
 
+    @GetMapping("/request-breakdown")
+    public RequestBreakdown requestBreakdown(
+            @RequestParam(required = false) @DateTimeFormat(iso = DateTimeFormat.ISO.DATE) @Nullable LocalDate dateFrom,
+            @RequestParam(required = false) @DateTimeFormat(iso = DateTimeFormat.ISO.DATE) @Nullable LocalDate dateTo) {
+        if (dateFrom != null && dateTo != null && dateFrom.isAfter(dateTo)) {
```
Comment (jescard-tamer-cecg) [inline api/service/src/main/java/com/coreeng/supportbot/prtracking/rest/TenantInsightsController.java]: **Inverted-date-range guard is now copy-pasted a 3rd time** (also in `prStats` and `escalationBreakdown`). Extract a small `validateDateRange(dateFrom, dateTo)` helper so the message and behavior stay consistent and the next endpoint doesn't add a 4th copy.

File: api/service/src/main/java/com/coreeng/supportbot/prtracking/JdbcPrTrackingRepository.java [hunk 7]
```diff
@@ -376,6 +376,51 @@ WHERE EXISTS (
...
+            binds.add(dateTo);
+        }
+        String sql = """
+                SELECT
+                    COUNT(DISTINCT t.id) AS total_support_tickets,
+                    COUNT(DISTINCT t.id) FILTER (
+                        WHERE EXISTS (SELECT 1 FROM pr_tracking pt WHERE pt.ticket_id = t.id)
+                    ) AS total_pr_tickets,
+                    COUNT(DISTINCT t.id) FILTER (
+                        WHERE EXISTS (SELECT 1 FROM pr_tracking pt WHERE pt.ticket_id = t.id)
```
Comment (jescard-tamer-cecg) [inline api/service/src/main/java/com/coreeng/supportbot/prtracking/JdbcPrTrackingRepository.java]: **Minor efficiency.** `EXISTS(pr_tracking …)` is evaluated twice — here and for `total_pr_tickets` above — since Postgres won't share the subexpression across two FILTER clauses. And the intervention `EXISTS(escalation … source = ?)` (line ~408) has no covering index (only the partial `opened` unique index + the FK). Cheap at current scale; if it matters later, compute a `has_pr` flag once and add an index on `escalation(ticket_id)` (which would also help `getEscalationBreakdown`).

File: api/service/src/main/java/com/coreeng/supportbot/prtracking/JdbcPrTrackingRepository.java [hunk 8]
```diff
@@ -376,6 +376,51 @@ WHERE EXISTS (
...
+                          AND EXISTS (
+                              SELECT 1 FROM escalation e
+                              WHERE e.ticket_id = t.id AND e.source = ?
```
Comment (jescard-tamer-cecg) [inline api/service/src/main/java/com/coreeng/supportbot/prtracking/JdbcPrTrackingRepository.java]: **Semantics worth confirming:** this counts a ticket as an intervention if *any* manual escalation row exists, regardless of whether it was later resolved (no status predicate). Probably intended — a human did intervene — and it's consistent with `getEscalationBreakdown`, but worth a sanity check against the 'Needed Manual Escalation' wording the funnel surfaces.

File: ui/src/components/tenant-requests/tenant-requests.tsx [hunk 9]
```diff
@@ -143,8 +143,17 @@ export default function TenantRequestsPage() {
 
   const repos = realRepos ?? [];
 
+  // Request funnel: total support requests → % that are PRs → % of those PRs needing intervention.
+  // All three counts share the ticket-creation date anchor, so they are nested subsets and the
+  // percentages are coherent (PR share never exceeds 100%).
+  const prPercentage =
```
Comment (jescard-tamer-cecg) [inline ui/src/components/tenant-requests/tenant-requests.tsx]: **Rounding can overstate the funnel.** `Math.round` on both percentages means 99.5–99.9% renders as 100% — e.g. 'Needed Manual Escalation' shows a warning-styled 100% when 199 of 200 PRs escalated, implying zero net automation. Cosmetic and consistent with the prior code, but consider `Math.floor` (or a decimal near the extremes) so a non-trivial remainder never rounds away.

File: ui/src/app/api/tenant-insights/request-breakdown/route.ts [hunk 10]
```diff
@@ -0,0 +1,27 @@
+import { backendFetch, errorResponse, unauthorizedResponse } from "../../_lib/backend-fetch";
```
Comment (jescard-tamer-cecg) [inline ui/src/app/api/tenant-insights/request-breakdown/route.ts:1]: Lowest priority: this handler is a near-verbatim copy of `escalation-breakdown/route.ts` (param parsing, `backendFetch`, error/502 handling). It follows the existing per-route convention, but it's now a 3rd copy — a shared `proxyTenantInsights(request, path)` helper would collapse them eventually.

File: api/service/src/main/java/com/coreeng/supportbot/prtracking/JdbcPrTrackingRepository.java [hunk 11]
```diff
@@ -376,6 +376,51 @@ WHERE EXISTS (
...
+        StringBuilder dateFilter = new StringBuilder();
+        if (dateFrom != null) {
+            dateFilter.append("AND q.date::date >= ?::date ");
+            binds.add(dateFrom);
+        }
+        if (dateTo != null) {
+            dateFilter.append("AND q.date::date <= ?::date ");
+            binds.add(dateTo);
+        }
+        String sql = """
```
Comment (jescard-tamer-cecg) [inline api/service/src/main/java/com/coreeng/supportbot/prtracking/JdbcPrTrackingRepository.java]: is there a reason here for using plain sql instead of j00q? 
it would make maintenance easier and cleaner code

File: api/service/src/main/java/com/coreeng/supportbot/prtracking/RequestBreakdown.java [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+        if (totalSupportTickets < 0 || totalPrTickets < 0 || interventionPrTickets < 0) {
+            throw new IllegalArgumentException("counts must not be negative");
```
Comment (jescard-tamer-cecg) [inline api/service/src/main/java/com/coreeng/supportbot/prtracking/RequestBreakdown.java]: can you please create a custom exception for each one here and handle them in a controller advice?

File: ui/src/components/tenant-requests/tenant-requests.tsx [hunk 13]
```diff
@@ -271,12 +280,54 @@ export default function TenantRequestsPage() {
           </TabsList>
 
           <TabsContent value="stats" className="space-y-6">
+            <div>
```
Comment (jescard-tamer-cecg) [inline ui/src/components/tenant-requests/tenant-requests.tsx:284]: could you please attach a screenhot for the changes here?

**Review by jescard-tamer-cecg — CHANGES_REQUESTED** [review]

File: api/service/src/main/java/com/coreeng/supportbot/prtracking/JdbcPrTrackingRepository.java (hunk 11 above)
Comment (SavvasM1) [reply api/service/src/main/java/com/coreeng/supportbot/prtracking/JdbcPrTrackingRepository.java]: yea I dont dissagree. Whole file is plain sql though , just going with convention I guess.

File: api/service/src/main/java/com/coreeng/supportbot/prtracking/RequestBreakdown.java (hunk 12 above)
Comment (SavvasM1) [reply api/service/src/main/java/com/coreeng/supportbot/prtracking/RequestBreakdown.java]: Ok will create custom exceptions

File: api/service/src/main/java/com/coreeng/supportbot/prtracking/PrDetectionService.java (hunk 1 above)
Comment (SavvasM1) [reply api/service/src/main/java/com/coreeng/supportbot/prtracking/PrDetectionService.java:902]: As discussed, this won't be addressed as it's the issue we are trying to solve. As a followup, we will have the overriden custom messages support PR grouping so everything is included.

File: api/service/src/main/java/com/coreeng/supportbot/prtracking/JdbcPrTrackingRepository.java (hunk 8 above)
Comment (SavvasM1) [reply api/service/src/main/java/com/coreeng/supportbot/prtracking/JdbcPrTrackingRepository.java]: Confirming intended.

**Review by jescard-tamer-cecg — DISMISSED** [review]

**Review by jescard-tamer-cecg — APPROVED** [review]

File: ui/src/components/tenant-requests/tenant-requests.tsx (hunk 13 above)
Comment (SavvasM1) [reply ui/src/components/tenant-requests/tenant-requests.tsx:284]: Added at the top :)
